docs/typescript.md

```diff
@@ -1,14 +1,597 @@
+  mainnet: '0x1',
+  sepolia: '0xaa36a7',
+} as const; // Type { readonly mainnet: '0x1'; readonly sepolia: '0xaa36a7'; }
+```
+
+```typescript
+type TransactionMeta = TransactionBase &
+  (
+    | {
+        status: Exclude<TransactionStatus, TransactionStatus.failed>;
+      }
+    | {
+        status: TransactionStatus.failed;
+        error: TransactionError;
+      }
+  );
+
+const updatedTransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected,
+};
+
+this.messagingSystem.publish(
+  `${controllerName}:transactionFinished`,
+  updatedTransactionMeta, // Expected type: 'TransactionMeta'
+);
+// Property 'error' is missing in type 'typeof updatedTransactionMeta' but required in type '{ status: TransactionStatus.failed; error: TransactionError; }'.ts(2345)
+```
+
+🚫 Add type annotation
+
+```typescript
+// Type 'TransactionMeta'
+const updatedTransactionMeta: TransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected,
+}; // resolves error
+```
+
+✅ Add `as const` and leave to inference
+
+```typescript
+// Type narrower than 'TransactionMeta': { status: TransactionStatus.rejected; ... }
+// Doesn't include 'error' property. Correctly narrowed between `TransactionMeta` discriminated union
+const updatedTransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected as const,
+}; // resolves error
+```
+
+### Type Annotations
+
+An explicit type annotation is acceptable for overriding an inferred type if...
+
+1. It can further narrow an inferred type, thus supplying type information that the compiler cannot infer or access.
+2. It is being used to enforce a type constraint rather than assign a type definition.
+
+##### Use the `satisfies` operator to enforce a type constraint or to validate the assigned type
+
+In the second case, an abstract, "narrowest supertype" is used to constrain or validate a type. TypeScript provides the `satisfies` operator for this use case. It is able to both enforce a type constraint and further narrow the assigned type through inference.
+
+🚫 Use a type annotation for type validation
+
+```typescript
+const updatedTransactionMeta: TransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected,
+};
+
+updatedTransactionMeta.error; // Property 'error' does not exist on type '{ status: TransactionStatus.approved | TransactionStatus.cancelled | TransactionStatus.confirmed | TransactionStatus.dropped | TransactionStatus.rejected | TransactionStatus.signed | TransactionStatus.submitted | TransactionStatus.unapproved; ... }'.(2339)
+```
+
+✅ Use the `satisfies` operator for type validation
+
+```typescript
+const updatedTransactionMeta = {
+  ...transactionMeta,
+  status: TransactionStatus.rejected as const,
+} satisfies TransactionMeta;
+
+updatedTransactionMeta.error; // Property 'error' does not exist on type '{ status: TransactionStatus.rejected; ... }'.(2339)
+```
+
+#### Acceptable usages of `:` annotations
```

Comment:
Nit: I know that type signatures are referenced as an acceptable use earlier in the document, but it might be worth briefly mentioning this again here as well, so we don't mistakenly create the opposite impression

Reply:
Removed this section: https://github.com/MetaMask/contributor-docs/pull/74/commits/f627cbcd3297d22d46a838cc2e4f2defdcc2f174, and moved content directly under "Type Annotation" heading: https://github.com/MetaMask/contributor-docs/pull/74/commits/37773897aad289d77c9f9a9f353a157525dcbc7a